src/Microsoft.AspNetCore.ResponseCompression/MimeTypeFilter.cs

```diff
@@ -0,0 +1,207 @@
+        public bool ShouldCompress(string mimeType)
+        {
+            bool? shouldCompress;
+
+            //check exact match (type/subtype)
+            shouldCompress = ShouldCompressExact(mimeType);
+
+            if (shouldCompress.HasValue)
+            {
+                return shouldCompress.Value;
+            }
+
+            //check partial match (type/*)
+            int slashPos = mimeType.IndexOf('/');
+
+            if (slashPos > -1)
+            {
+                string wildcardMimeType = $"{mimeType.Substring(0, slashPos + 1)}*";
+                shouldCompress = ShouldCompressExact(wildcardMimeType);
+
+                if (shouldCompress.HasValue)
+                {
+                    return shouldCompress.Value;
+                }
+            }
+
+            //check wildcard (*/*)
+            shouldCompress = ShouldCompressExact("*/*");
+
+            if (shouldCompress.HasValue)
+            {
+                return shouldCompress.Value;
+            }
+
+            //no matches and no "*/*" entry in either list - do not compress
+            return false;
+        }
+
+        private MimeTypeFilter Add(ISet<string> set, IEnumerable<string> mimeTypes)
+        {
+            if (mimeTypes != null)
+            {
+                foreach (string mimeType in mimeTypes)
+                {
+                    if (!IsValid(mimeType))
+                    {
+                        throw new ArgumentException(
+                            $@"MIME type ""{mimeType}"" does not match the format ""<type>/<sybtype>"", ""<type>/*"" or ""*/*"".",
+                            nameof(mimeTypes)
+                        );
+                    }
+
+                    set.Add(mimeType);
+                }
+            }
+
+            HasBeenPopulated = true;
+            return this;
+        }
+
+        private MimeTypeFilter Remove(ISet<string> set, IEnumerable<string> mimeTypes)
+        {
+            if (mimeTypes != null)
+            {
+                set.ExceptWith(mimeTypes);
+            }
+
+            HasBeenPopulated = true;
+            return this;
+        }
+
+        private MimeTypeFilter Clear(ISet<string> set)
+        {
+            set.Clear();
+            HasBeenPopulated = true;
+            return this;
+        }
+
+        //This is NOT a proper MIME type validation. It only checks conformance to the format
+        //"type/sybtype", "type/*" or "*/*", for the purpose of being used in the lists of MIME
+        //types that should and should not be compressed.
+        private static bool IsValid(string mimeType) =>
+            Regex.IsMatch(mimeType, @"^([^\s\*/]+/([^\s\*/]+|\*)|\*/\*)$", RegexOptions.Compiled);
+
+        private bool? ShouldCompressExact(string mimeType)
+        {
+            //check not compressed first, then compressed
+            if (_notCompressed.Contains(mimeType))
```

Comment:
With this approach a notCompressed wildcard (image/*) would loose vrs an exact _compressed entry (image/foo). That makes wildcards in notCompressed ineffective and you might as well not check for it.

Reply:
Can I confirm what you mean? Are you saying that if I checked `_compressed` for the exact match and it's not there, there's no point checking `_notCompressed` for partial match, because the end result will be the same?

Consider the following setup: `_compressed = ["image/foo", "*/*"]`, `_notCompressed = ["image/*"]`. If I'm testing `image/bar` and check `_notCompressed` for partial match, `SouldCompress` will return `false`. But if I don't check `_notCompressed` for partial match, it'll return `true` because of the `*/*` entry in `_compressed`.

Reply:
Ah, I hadn't considered the `*/*` case.